### PR TITLE
Upgrade d3 geo projections

### PR DESCRIPTION
**Pull request by archmoj**
Replace old `d3.geo` functions with upgraded `d3-geo` modules (that support IE) as part of `d3 bump` cc: #424.
Some new `geo` projections e.g. `satelite` could potentially be added as listed in https://github.com/plotly/plotly.js/pull/5112/commits/8dd4ba0658e5c93279816be187688eba1ba4b04e to address #4781.

This would also result in reducing the size of various partial bundles that do not need `geo` trace.

@plotly/plotly_js

### CODE REVIEW

**Comment by alexcjohnson** [comment]
This looks like a good idea, but I think we need to be more careful about each projection we add: (1) does it work well in the simple case? In your demo codepen a number of them fall look pretty bad either immediately or easily as you zoom / pan around so should probably not be included without some adjustment. (2) maybe you've already done this, do they support all the same parameters as the existing ones? `isConic`? `lonaxisSpan` / `lataxisSpan`? Any other special cases? (3) each one we add needs to be included in at least one mock, maybe more so we can ensure their parameters work right. `geo_fitbounds-locations` says it includes "all projection types", let's ensure that stays true.

File: package.json [hunk 1]
```diff
@@ -57,7 +57,7 @@
     ]
   },
   "dependencies": {
-    "@plotly/d3": "^3.5.18",
+    "@plotly/d3": "git://github.com/plotly/d3.git#e792de9ae52d956d18582cac8a49ea6a03c7e010",
```
Comment (archmoj) [inline package.json:60]: https://github.com/plotly/d3/compare/master...plotly:drop-geo

**Comment by archmoj** [comment]
> This looks like a good idea, but I think we need to be more careful about each projection we add: (1) does it work well in the simple case? In your demo codepen a number of them fall look pretty bad either immediately or easily as you zoom / pan around so should probably not be included without some adjustment. (2) maybe you've already done this, do they support all the same parameters as the existing ones? `isConic`? `lonaxisSpan` / `lataxisSpan`? Any other special cases? (3) each one we add needs to be included in at least one mock, maybe more so we can ensure their parameters work right. `geo_fitbounds-locations` says it includes "all projection types", let's ensure that stays true.

Thanks for the notes @alexcjohnson.
With the updates (reflected in the PR description now) no new projections were added in the API as a result of this PR.

File: test/jasmine/tests/geo_test.js [hunk 2]
```diff
@@ -1353,46 +1353,6 @@ describe('Test geo interactions', function() {
         .then(done, done.fail);
     });
 
-    it('should clear hover label when cursor slips off subplot', function(done) {
```
Comment (alexcjohnson) [inline test/jasmine/tests/geo_test.js:1356]: Why did these 2 tests get removed?

Comment (archmoj) [reply test/jasmine/tests/geo_test.js:1356]: If I recall (as mentioned in the commit message) there used to be a bug in d3.geo where a function returned `NaN` positions. And we guarded against it.

Comment (alexcjohnson) [reply test/jasmine/tests/geo_test.js:1356]: But these tests are just using the high-level API, not d3 geo itself. If they started failing, that means something changed about the user-facing behavior. What changed? Are we comfortable with that change? I'd much prefer if we can adapt the tests to the new version so we can see the difference.

**Comment by nicolaskruchten** [comment]
We don't strictly speaking need to merge this PR for v2, so let's leave it alone for now?

File: test/jasmine/tests/geo_test.js (hunk 2 above)
Comment (archmoj) [reply test/jasmine/tests/geo_test.js:1356]: Good question. Pointing to the fact that rebase somehow forgot to remove `geo/projections`. 
It is now fixed in https://github.com/plotly/plotly.js/pull/5112/commits/557ebde74e01c6df45c59083fc7c570a3e21ad70.

**Comment by archmoj** [comment]
> We don't strictly speaking need to merge this PR for v2, so let's leave it alone for now?

That's what we may end up doing. But it is worth having the conversion going before committing to v2 IMHO.

File: test/jasmine/tests/geo_test.js (hunk 2 above)
Comment (alexcjohnson) [reply test/jasmine/tests/geo_test.js:1356]: This test should still be able to pass though. Perhaps the exact pixel positions need to change because there's some small change in how the projection is applied? If I try to manually follow what's going on in the test (I believe it's hovering on the point off the coast of Africa, then slowly mousing off the edge of the globe) the behavior looks identical before/after the change. So we should be able to bring the test back, but if the exact values need to change a little (`px = 390`, `py = 290` and `if(px < 402)` that's fine.

File: test/jasmine/tests/geo_test.js [hunk 3]
```diff
@@ -1421,66 +1381,6 @@ describe('Test geo interactions', function() {
...
-                lataxis: {
-                    range: [38.794799, 45.122292]
-                },
-                lonaxis: {
-                    range: [-82.904731, -81.622334]
-                }
-            },
-            width: 700,
-            heigth: 500
-        })
```
Comment (alexcjohnson) [inline test/jasmine/tests/geo_test.js:1454]: This case looks like it's actually broken in this PR. If I create the above plot on `latest` I get:

Screen Shot 2021-01-20 at 12 51 26 PM

But on this branch I get no warning and a completely blank plot.

File: test/jasmine/tests/geo_test.js (hunk 2 above)
Comment (archmoj) [reply test/jasmine/tests/geo_test.js]: Good call. Adjusted in https://github.com/plotly/plotly.js/pull/5112/commits/647b0dc5576e2b57c1b3425bb152a631a24d3e06.
In fact with previous positions the tests didn't pass locally. But now the test now passes both locally and on the CI.

File: test/jasmine/tests/geo_test.js (hunk 3 above)
Comment (archmoj) [reply test/jasmine/tests/geo_test.js:1454]: The default view is the fallback resulted from catching `d3-v3.geo` error.
We get a blank view here as we are zoomed into the graph with that particular setup.
On could zoom out by clicking on the modebar.
So it looks like we could remove the fallback and the related test by https://github.com/plotly/plotly.js/pull/5112/commits/a046871d6931e6c8f33a5441c6e6ffc385e114b9.

File: src/plots/geo/geo.js [hunk 4]
```diff
@@ -3,6 +3,10 @@
 /* global PlotlyGeoAssets:false */
 
 var d3 = require('@plotly/d3');
+var geo = require('d3-geo');
+var geoPath = require('d3-geo').geoPath;
+var geoDistance = require('d3-geo').geoDistance;
```
Comment (alexcjohnson) [inline src/plots/geo/geo.js:8]:
```suggestion
var geoPath = geo.geoPath;
var geoDistance = geo.geoDistance;
```

File: test/jasmine/tests/geo_test.js (hunk 3 above)
Comment (alexcjohnson) [reply test/jasmine/tests/geo_test.js:1454]: OK, I guess this makes sense. The map is still in a somewhat broken state at that point: if you scroll-zoom until you can see the map appear from above, you can drag it up and down but not left/right. If you then (or instead) use the zoom out button, now you can drag left/right. Can we detect and fix this in the scroll zoom handler perhaps?

**Comment by archmoj** [comment]
Conflicts are resolved here and the plotly/d3 PR.
Now adding this to [v2.3.0 milestone](https://github.com/plotly/plotly.js/milestone/63).

File: test/jasmine/tests/geo_test.js (hunk 3 above)
Comment (archmoj) [reply test/jasmine/tests/geo_test.js:1454]: I created an issue (see #5787) to track this edge case so that we could move on.

**Review by alexcjohnson — APPROVED** [review]
Looks good, thanks for making the new issue.
💃 after publishing the updated `@plotly/d3`

File: package.json [hunk 5]
```diff
@@ -62,7 +62,7 @@
     ]
   },
   "dependencies": {
-    "@plotly/d3": "^3.6.1",
+    "@plotly/d3": "git://github.com/plotly/d3.git#d4fef4759a7d7dd29360cbb891ca53422f9f689e",
```
Comment (archmoj) [inline package.json:65]: https://github.com/plotly/d3/pull/3
